### PR TITLE
Feature/modbus data type

### PR DESCRIPTION
**Pull request by hutcheb**
Add support for BOOL, INT, UINT and REAL data types for the Modbus protocol.
address string can now include any of these types such as:- holding-register:10:REAL[10], extended-register:10:UINT[10].
Note the previous format now defaults to BOOL or INT depending on the memory area.

Add support to parse values from Java Strings to any of these types. This fixes a bug in the hello-world example when writing.

Add support for BOOL, INT, UINT and REAL PlcValue types. This have been added as we are able to include the max/min value checks and casting of the input type within the constructors cleaning up the ModbusFieldHandler logic. The method getBytes has also been added to clean up the ModbusProtocolLogic class.

I'll add Documentation once the other Modbus Documentation PR has been finalized.

### CODE REVIEW

**Comment by niclash** [comment]
I have also seen 32-bit integers. Maybe it should be available too.

**Comment by hutcheb** [comment]
Hi Niclas,

The plan is to add support for as many IEC 61131 data types as possible. I've added DINT as an example for what it would take to add additional data types but will wait for more feedback before adding more.

Ben

File: plc4j/api/src/main/java/org/apache/plc4x/java/api/value/PlcBOOL.java [hunk 1]
```diff
@@ -0,0 +1,214 @@
...
+import org.apache.plc4x.java.api.value.*;
+
+import java.math.BigDecimal;
+import java.math.BigInteger;
+
+@JsonTypeInfo(use = JsonTypeInfo.Id.CLASS, property = "className")
+public class PlcBOOL extends PlcIECValue<Boolean> {
+
+    BigInteger minValue = BigInteger.valueOf(0);
+    BigInteger maxValue = BigInteger.valueOf(1);
```
Comment (chrisdutz) [inline plc4j/api/src/main/java/org/apache/plc4x/java/api/value/PlcBOOL.java]: Could we perhaps make these static ... A BigInteger is a pretty big type and creating these for every boolean PlcValue seems wasting memory and CPU time for cleaning up.

File: plc4j/api/src/main/java/org/apache/plc4x/java/api/value/PlcDINT.java [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+import java.math.BigDecimal;
+import java.math.BigInteger;
+import java.util.LinkedList;
+import java.util.List;
+
+@JsonTypeInfo(use = JsonTypeInfo.Id.CLASS, property = "className")
+public class PlcDINT extends PlcIECValue<Integer> {
+
+    BigInteger minValue = BigInteger.valueOf((long) Integer.MIN_VALUE);
+    BigInteger maxValue = BigInteger.valueOf((long) Integer.MAX_VALUE);
```
Comment (chrisdutz) [inline plc4j/api/src/main/java/org/apache/plc4x/java/api/value/PlcDINT.java]: Same as with PlcBOOL (And probably all others)

**Review by chrisdutz — APPROVED** [review]
I really like this PR ... I guess it would simplify driver implementation quite a bit ... the thing with the BigIntegers as instance variables we can always change.

**Comment by chrisdutz** [comment]
Do you need any help from my side in order to continue? If yes, just reach out to me.

**Comment by hutcheb** [comment]
I have created all of the numeric datatypes and char and wchar types as well as updated the ModbusFieldHandler file.

I'm just sorting out some issues this morning with how BigIntegers are converted to byte[] within the PlcULINT and PlcLWORD classes.

By the end of my session this morning I should be in a position to start writing test cases in the Modbus section.

I haven't touched the other drivers.

I'll commit what I have when I've finished this morning next 15mins or so, feel free to carry on as I won't be touching it until tomorrow.

**Comment by hutcheb** [comment]
Hi Chris,

This should be in a position for you to review.

Ben

**Comment by chrisdutz** [comment]
Hi Ben,

will do so asap …

Chris

Von: hutcheb <notifications@github.com>
Antworten an: apache/plc4x <reply@reply.github.com>
Datum: Samstag, 19. September 2020 um 13:30
An: apache/plc4x <plc4x@noreply.github.com>
Cc: Christofer Dutz <christofer.dutz@c-ware.de>, Comment <comment@noreply.github.com>
Betreff: Re: [apache/plc4x] Feature/modbus data type (#186)


Hi Chris,

This should be in a position for you to review.

Ben

**Comment by hutcheb** [comment]
Hi,

I was looking at starting to use the staticSerialize method for Modbus as it would reduce the fromPLCValue method. However there are a few methods in the WriteBuffer class that aren't implemented. writeDouble and the writeBigInteger functions. The writeDouble function in particular would cause issues with writing floats of any kind for all protocols that use it.

Does writing floats with the S7 protocol work or am I missing something?

**Comment by chrisdutz** [comment]
Hi Ben ... sorry for the late response ... I'm currently drowning in getting some stuff finished as well as preparing for ApacheCon next week ... I'll try to squeeze it in, but can't promise ... haven't forgotten you and your hard work ... just want to invest the time for reviewing, that your work deserves.

**Comment by hutcheb** [comment]
Hi Chris, not a problem take your time. :)

**Comment by chrisdutz** [comment]
The PR #190 now contains all changes of this PR.

**Pull request closed without merging**